### PR TITLE
impelements OIDC kubeconfig generation

### LINKED ISSUE #2268: Implement API EP that generates kubeconfig with OIDC tokens
**User Story** TBD

**Acceptance criteria** TBD

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: adds a new HTTP endpoint `v1/kubeconfig` for generating `kubeconfig` for the given cluster. The config will contain OIDC tokens. The endpoint is optional and can be enabled by passing additional command line flags. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2268 

**Release note**:

```release-note
Added a new HTTP endpoint for generating kubeconfig with OIDC tokens. The path of the endpoint is `api/v1/kubeconfig`

[Action required] users need to add redirection to api/v1/kubeconfig to dex installation
[Action required] kubermatic API command line args were changed:`token-issuer` to `oidc-token-issuer`, `token-issuer-skip-tls-verify` to `oidc-skip-tls-verify`, `client-id` to `oidc-client-id`
[Action required] to enable the endpoint the users must set `-feature-gates=OIDCKubeCfgEndpoint=true` and provide `-oidc-client-secret`, `-oidc-redirect-uri` when running the server. If the provider they are working with doesn't accept `refresh_token` as a scope, setting `-oidc-offline-access-as-scope=false` will pass appropriate query parameter instead.
```

### CODE REVIEW

File: api/pkg/handler/routes_v1_optional.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			r.userInfoMiddlewareUnauthorized(),
+		)(createOIDCKubeconfig(r.projectProvider, r.oidcIssuer, oidcCfg)),
+		decodeCreateOIDCKubeconfig,
+		encodeKubeconfigDoINeddAcditional,
+		r.defaultServerOptions()...,
+	)
+}
+
+// Remove this EP
+func (r Routing) redirectTo(path string) http.Handler {
```
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1_optional.go]: Will remove once https://github.com/kubermatic/secrets/pull/178 is merged.

File: api/pkg/handler/kubeconfig.go [hunk 2]
```diff
@@ -30,6 +36,125 @@ func getClusterKubeconfig(projectProvider provider.ProjectProvider) endpoint.End
...
+
+		// PHASE initial handles request from and end-user that want to authenticate
+		// and kickoff the process of kubeconfig generation
+		if req.phase != initialPhase {
+			return nil, kcerrors.NewBadRequest("bad request unexpected ")
+		}
+
+		// TODO: pass nonce
+		rsp := createOIDCKubeconfigRsp{}
+		// TODO: Define a proper list of scopes, is offline_access only a scope ? (you can set it also via clientlib)
```
Comment (p0lyn0mial) [inline api/pkg/handler/kubeconfig.go]: It seems like for some providers `offline_access` can be also passed as an option https://github.com/dexidp/dex/blob/master/vendor/golang.org/x/oauth2/oauth2.go#L122:18

Comment (p0lyn0mial) [reply api/pkg/handler/kubeconfig.go]: for `Google` passing as scope works just fine. Not sure about `LDAP`

**Comment by p0lyn0mial** [comment]
the endpoint generates the following kubeconfig
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURHakNDQWdLZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREErTVR3d09nWURWUVFERXpOeWIyOTAKTFdOaExuRjNhMlp1YXpWaWVIZ3VaWFZ5YjNCbExYZGxjM1F6TFdNdVpHVjJMbXQxWW1WeWJXRjBhV011YVc4dwpIaGNOTVRneE1ETXhNVE0wTkRNM1doY05Namd4TURJNE1UTTBORE0zV2pBK01Ud3dPZ1lEVlFRREV6TnliMjkwCkxXTmhMbkYzYTJadWF6VmllSGd1WlhWeWIzQmxMWGRsYzNRekxXTXVaR1YyTG10MVltVnliV0YwYVdNdWFXOHcKZ2dFaU1BMEdDU3FHU0liM0RRRUJBUVVBQTRJQkR3QXdnZ0VLQW9JQkFRRFF2NlJpZDNVa3hIRGlHT0ZpbFRmOQpBS3hXdTN1YkhxSTRSZi9YU0o5MnU2RXA0blAyS0R4K0RxYWpINzRFbkNsYWxoeWRYUFAxZ084SXpabVdyUURtCkJRYTRUdEx0cnRTK2JxM2ZIK3pDV1NMSjVhS1o5Z3J3eUwrUlZUUnJzQ2hzR20wQWt5cW9YSkZKQ244WmRjREcKMm5QOU1FSkd3L1VEYUp5dVRZSUtSTGNEZ0VIc2IvT0I1WkI1a1VKMFBGdURrdHFJSGNhY24xTTNwL0tsMEllQwp3ZTNWQVViRWw0WnB2WWxrUEJtNFNKODRjZm1Sai9QOEpYNlk4TkZsRC9ZaUx6QWt6dGp5YkloWUwrMjFCTHhQCnlXRFd4aGt4dldUa0JZNjJuak4vSFErYWJrUFA5bUdwT21CWUlGQ2gxUkpXd0IyMTlkdkx3bnNqUlJ2cTVVczUKQWdNQkFBR2pJekFoTUE0R0ExVWREd0VCL3dRRUF3SUNwREFQQmdOVkhSTUJBZjhFQlRBREFRSC9NQTBHQ1NxRwpTSWIzRFFFQkN3VUFBNElCQVFDUjZNUjFjTDR2SEtSVWpsN0w0TldaSVpNcmVtSlJqb2pTVCtVSWtyTDVoalF1CmgwTmJJSkIxdkpQS0J6RThTYVdySEJNMHpsRFFpZ1Z3azJZQW9EVmlmU0ZPeEt3S0JoZ2c4WlluYW9kb2dqZngKeU5WK240Z0lZMFdtTWMvNEhJT2l1NFUxYXRNK2R1L1lTbkFiSlp6NlhrTVphR1laSWpYL3Y2WU9ZMkt5WE10YwpRZ05zcy9jZFQwbjg2S2w1cWwzTW5vREk4akRiZnQ4by9EU3UzSk5jQ3Zsb3FLdzNpVVdwNVVUQnAvTGpEamgwCi8xdFpGZGVRcFRta1JTNlZqRm9LQ0E1bHRZemhuejIxN0JwY0NSQ21MSFQrV3p4aFExZ1lnaEJOVHYwWi91bXgKREo0T2greHdvY1V6WXprcHBlT3o2Rys3N3JDckx6OFhmWWpYM3RaLwotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
    server: https://qwkfnk5bxx.europe-west3-c.dev.kubermatic.io:31056
  name: qwkfnk5bxx
contexts:
- context:
    cluster: qwkfnk5bxx
    user: lukasz@loodse.com
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: lukasz@loodse.com
  user:
    auth-provider:
      config:
        client-id: REDACTED
        client-secret: REDACTED
        id-token: REDACTED
        idp-issuer-url: https://dev.kubermatic.io/dex
        refresh-token: REDACTED
      name: oidc
```


haven't tested if this works.

**Comment by p0lyn0mial** [comment]
@mrIncompetent ready for review, PTAL.

File: api/pkg/handler/kubeconfig.go [hunk 3]
```diff
@@ -30,6 +36,129 @@ func getClusterKubeconfig(projectProvider provider.ProjectProvider) endpoint.End
...
+			rsp := createOIDCKubeconfigRsp{}
+			rsp.phase = kubeconfigGenerated
+			rsp.oidcKubeConfig = oidcKubeCfg
+			return rsp, nil
+		}
+
+		// PHASE initial handles request from and end-user that want to authenticate
+		// and kickoff the process of kubeconfig generation
+		if req.phase != initialPhase {
+			return nil, kcerrors.NewBadRequest("bad request unexpected ")
```
Comment (mrIncompetent) [inline api/pkg/handler/kubeconfig.go]: Please set a clear error message

File: api/pkg/handler/routing_test.go [hunk 4]
```diff
@@ -13,6 +12,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/go-test/deep"
+
```
Comment (mrIncompetent) [inline api/pkg/handler/routing_test.go]: please remove the empty line here.

**Comment by p0lyn0mial** [comment]
comments addressed, thanks for the review, ptal.

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**